pkg/kubernetes/kubernetes.go

```diff
@@ -40,57 +38,46 @@ func New(s v1alpha1.Spec) *Kubernetes {
 	return &k
 }
 
-// Manifest describes a single Kubernetes manifest
-type Manifest map[string]interface{}
-
-func (m Manifest) Kind() string {
-	return m["kind"].(string)
-}
-
-func (m Manifest) Name() string {
-	return m["metadata"].(map[string]interface{})["name"].(string)
-}
-
-func (m Manifest) Namespace() string {
-	return m["metadata"].(map[string]interface{})["namespace"].(string)
-}
-
-// Reconcile receives the raw evaluated jsonnet as a marshaled json dict and
-// shall return it reconciled as a state object of the target system
-func (k *Kubernetes) Reconcile(raw map[string]interface{}, objectspecs []*regexp.Regexp) (state []Manifest, err error) {
-	docs, err := walkJSON(raw, "")
-	out := make([]Manifest, 0, len(docs))
-	if err != nil {
-		return nil, errors.Wrap(err, "flattening manifests")
-	}
-	for _, d := range docs {
-		m := objx.New(d)
-		if k != nil && !m.Has("metadata.namespace") {
-			m.Set("metadata.namespace", k.Spec.Namespace)
-		}
-		out = append(out, Manifest(m))
-	}
-
-	if len(objectspecs) > 0 {
-		out = funk.Filter(out, func(i interface{}) bool {
-			p := objectspec(i.(Manifest))
-			for _, o := range objectspecs {
-				if o.MatchString(strings.ToLower(p)) {
+// Build takes the raw evaluated jsonnet as a marshaled json dict, along with some
+// <kind>/<name> matching regular expressions. It will return a list of Manifests, which
+// it finds by recursively walking the jsonnet structure, and filtering each manifest
+// against the matchers.
+//
+// In addition, we sort the manifests to ensure the order is consistent in each
+// show/diff/apply cycle. This isn't necessary, but it does help users by producing
+// consistent diffs.
+func (k *Kubernetes) Build(raw map[string]interface{}, kindNameMatchers []*regexp.Regexp) []Manifest {
```

Comment:
> Actually, having thought about this, I think it's a really bad idea for us to silently fail if bad manifests are in the jsonnet output. I'm going to adjust walkJSON to fail if this is the case, as it did before.

As you have figured out, this makes debugging very hard and guessy. IMO Tanka should expect users to only ever output valid objects. Otherwise kubernetes objects might end up at invalid locations / the other way around and the user does not immediately recognize there even is an error .. only his object does not show up